pliers/extractors/text.py

```diff
@@ -402,3 +405,131 @@ def _extract(self, stim):
+        self.tokenizer_type = tokenizer
+        self.framework = framework
+        self.encoding_level = encoding_level
+        self.pooling = pooling
+        model_name = 'BertModel' if self.framework == 'pt' else 'TFBertModel'
+
+        self.tokenizer = transformers.BertTokenizer.from_pretrained(
+            tokenizer, **model_kwargs)
+        self.model = getattr(transformers, model_name).from_pretrained(
+            pretrained_model_or_path, **tokenizer_kwargs)
+
+        super(PretrainedBertEncodingExtractor, self).__init__()
+
+    def _extract(self, stims):
+
+        text, onsets, durations = zip(
+            *((s.text, s.onset, s.duration) for s in stims.elements))
+        durations = [0.0 if d is None else d for d in durations]
+
+        tokens = [self.tokenizer.tokenize(t) for t in text]
+        tokens_flat = list(flatten(tokens))
+        token_positions = np.arange(0, len(tokens_flat))
+
+        def cast_to_token_level(word_level_list):
+            token_level_list = [listify(word_level_list[i]) * len(tok)
+                                for i, tok in enumerate(tokens)]
+            token_level_list = list(flatten(token_level_list))
+            return token_level_list
+
+        t_text, t_ons, t_dur = map(cast_to_token_level,
+                                   [text, onsets, durations])
+
+        tensor_tokens = self.tokenizer.encode(tokens_flat,
+                                              return_tensors=self.framework)
+        output = self.model(tensor_tokens)
+        output = [out.detach().numpy() if self.framework == 'pt' else
+                  out.numpy() for out in output]
+
+        if self.encoding_level == 'token':
+            encoded_tokens = tokens_flat
+            encodings = output[0][:, 1:-1, :].squeeze()
+
+        elif self.encoding_level == 'sequence':
+            encoded_tokens = [' '.join(text)]
+            token_positions = ['None']
+            t_text = ['None']
+            t_dur = t_ons[-1] + t_dur[-1] - t_ons[0]
+            t_ons = t_ons[0]
+
+            if self.pooling:
+                pooling_function = getattr(np, self.pooling)
+                encodings = pooling_function(output[0][:, 1:-1, :],
+                                             axis=1, keepdims=True)
+            else:
+                encodings = output[1]
+
+        def expand_metadata(metadata):
+            return listify(metadata) * len(encoded_tokens)
+
+        seq, model, tokenizer, pooling = map(expand_metadata,
+                                             [' '.join(text),
+                                              self.pretrained_model,
+                                              self.tokenizer_type,
+                                              str(self.pooling)])
+
+        data = [encodings.tolist(), encoded_tokens, t_text, token_positions,
+                seq, model, tokenizer, pooling]
+        features = ['encoding', 'token', 'word', 'token_position', 'sequence',
```

Comment:
I think some of these should not be features, per se. 

For sure, I think `model`, `tokenizer` and `pooling` should just be attributes of the extractor rather than features. These can always be retrieved from the `Extractor` model. 

Also, shouldn't the pliers built-in `order` be used instead of `token_position`?

Otherwise, the rest seem fine, as they have to do with tokenization done by BERT, which is important.

Reply:
fixed that in latest commit by dropping those as `features `and adding some `log_attributes` to result df if specified in BERT-specific `to_df` method.
Wondering if it'd be good to add this to generic `to_df` method though (@tyarkoni)

Reply:
The generic `to_df` method already includes the `_log_attributes` dictionary as a field, I believe. Beyond that, adding options to `to_df` to control its behavior is complicated by the fact that it's almost always invoked implicitly (from the `ExtractorResult` initializer). So making use of it would require passing through an argument from the `_extract` call in every extractor, which seems clunky.

Reply:
> The generic `to_df` method already includes the `_log_attributes` dictionary as a field, I believe. </br>

I guess the result object does, so definitely always possible to retrieve that info. 

> Beyond that, adding options to `to_df` to control its behavior is complicated by the fact that it's almost always invoked implicitly (from the `ExtractorResult` initializer). So making use of it would require passing through an argument from the `_extract` call in every extractor, which seems clunky. </br>

Ok, still leaving this logic for the bert-specific `to_df` for now though, unless you suggest avoiding to do so here too?

Reply:
> Ok, still leaving this logic for the bert-specific to_df for now though, unless you suggest avoiding to do so here too?

If you can already get it from the standard handling, I think you can remove it here. If there's information that's *not* available from the DF otherwise, then leave it in. Either way, I don't think there's a need for the `model_attributes` argument, as I don't think that can currently actually be manipulated by the user via the public API (i.e., you'd have to call `_to_df()` directly, which is discouraged).

Reply:
> If you can already get it from the standard handling, I think you can remove it here. If there's information that's _not_ available from the DF otherwise, then leave it in. Either way, I don't think there's a need for the `model_attributes` argument, as I don't think that can currently actually be manipulated by the user via the public API (i.e., you'd have to call `_to_df()` directly, which is discouraged). </br>

Looks like you _can_ pass `model_attributes` to `to_df` and it will handle them correctly, as `to_df` takes kwargs and passes them to `_to_df()` when calling it internally. 

I can't find a way to plug `log_attributes` into result dataframe from generic `to_df` method though, and find no reference to `log_attributes` where `to_df` is defined. 
Caveat: it might be me not knowing pliers well enough yet.